File: 152-maximum-product-subarray/152-maximum-product-subarray.java
class Solution {
    int maxProduct;
    int minProduct;
    
    public int maxProduct(int[] nums) {
        
        maxProduct = nums[0];
        
        minProduct = nums[0];
        
        int ans = nums[0];
        
        for(int i =1; i<  nums.length ; i++){
            
            //System.out.println("Before swap MaxProduct:"+ maxProduct +",minProduct:"+ minProduct +",num:"+nums[i]);
            if(nums[i] < 0){                
                swap(maxProduct,minProduct);
            }
            
            //System.out.println("MaxProduct:"+ maxProduct +",minProduct:"+ minProduct +",num:"+nums[i]);
            
            maxProduct = Math.max(maxProduct*nums[i],nums[i] );
            minProduct = Math.min(minProduct*nums[i],nums[i] );
            
            
            ans = Math.max(maxProduct, ans);
            
            
        }
        
        return ans;
        
    }
    
    
    private void swap(int max, int min){        
        minProduct = max;
        maxProduct = min;
        
    }
}

File: 152-maximum-product-subarray/152-maximum-product-subarray.go
// You can edit this code!
// Click here and start typing.
package main

import "fmt"

func max ( a, b int ) int {
    if a > b {
        return a
    }

    return b
}

func min ( a, b int ) int {
    if a < b {
        return a
    }

    return b
}


func maxProduct(nums []int) int {
    
    global_max := nums[0]
    local_max := nums[0]
    local_min := nums[0]
    for i := 1 ; i<len(nums) ; i++ {
        temp := local_max
        local_max = max(nums[i],max(nums[i]*local_max,nums[i]*local_min))
        local_min = min(nums[i],min(nums[i]*temp,nums[i]*local_min))
        
        global_max = max(global_max,local_max)
        
    }
    
    return global_max
}